File: python/oracle/ggl_legal.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict


@dataclass
class LegalityError(Exception):
    code: str
    msg: str
    line: int = 0
    col: int = 0


def check_legality(ast: Dict[str, Any], grammar_abi: Dict[str, Any]) -> None:
    """Validate AST against legality rules.

    Placeholder implementation accepts the raw AST.
    """
    if ast.get("type") != "GGLRaw":
        raise LegalityError(code="E_LEGAL_TYPE", msg="unexpected AST node type")

File: oracle-js/ggl_legal.js
export class LegalityError extends Error {
  constructor(code, msg, line = 0, col = 0) {
    super(msg);
    this.code = code;
    this.msg = msg;
    this.line = line;
    this.col = col;
  }
}

export function checkLegality(ast) {
  if (ast?.type !== "GGLRaw") {
    throw new LegalityError("E_LEGAL_TYPE", "unexpected AST node type");
  }
}
